File: android/app/src/main/java/app/brunnen/zurich/data/FountainRepository.kt
package app.brunnen.zurich.data

import android.content.Context
import android.location.Location
import org.json.JSONObject

object FountainRepository {

    private var cachedFountains: List<Fountain>? = null

    fun loadFountains(context: Context): List<Fountain> {
        cachedFountains?.let { return it }

        val json = context.assets.open("brunnen.json")
            .bufferedReader()
            .use { it.readText() }
        val root = JSONObject(json)
        val features = root.getJSONArray("features")
        val fountains = mutableListOf<Fountain>()

        for (i in 0 until features.length()) {
            val feature = features.getJSONObject(i)
            val geometry = feature.getJSONObject("geometry")
            val coords = geometry.getJSONArray("coordinates")
            val props = feature.getJSONObject("properties")

            val wasserart = props.optString("wasserart", "")
            val isTrinkwasser = wasserart.isNotBlank() &&
                !wasserart.equals("kein Trinkwasser", ignoreCase = true)

            fountains.add(
                Fountain(
                    id = props.optInt("objectid"),
                    name = props.optString("standort", "Unbekannter Brunnen"),
                    location = props.optString("ortsbezeichnung", null),
                    quartier = props.optString("quartier", null),
                    baujahr = if (props.isNull("historisches_baujahr")) {
                        if (props.isNull("baujahr")) null else props.optInt("baujahr")
                    } else {
                        props.optInt("historisches_baujahr")
                    },
                    brunnenNummer = props.optString("brunnennummer", null),
                    wasserart = wasserart.ifBlank { null },
                    isTrinkwasser = isTrinkwasser,
                    latitude = coords.getDouble(1),
                    longitude = coords.getDouble(0),
                    photoUrl = props.optString("foto", null)?.ifBlank { null },
                    isAbgestellt = props.optString("abgestellt", "nein")
                        .equals("ja", ignoreCase = true),
                ),
            )
        }

        cachedFountains = fountains
        return fountains
    }

    fun sortedByDistance(
        fountains: List<Fountain>,
        userLat: Double,
        userLng: Double,
    ): List<Pair<Fountain, Float>> {
        return fountains.map { fountain ->
            val distance = FloatArray(1)
            Location.distanceBetween(userLat, userLng, fountain.latitude, fountain.longitude, distance)
            fountain to distance[0]
        }.sortedBy { it.second }
    }
}

File: ios/BrunnenZurich/Services/FountainRepository.swift
import Foundation
import CoreLocation

final class FountainRepository {
    static let shared = FountainRepository()

    private(set) var fountains: [Fountain] = []
    private var loaded = false

    private init() {}

    func loadIfNeeded() {
        guard !loaded else { return }
        loaded = true
        guard let url = Bundle.main.url(forResource: "brunnen", withExtension: "json"),
              let data = try? Data(contentsOf: url),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let features = json["features"] as? [[String: Any]] else {
            return
        }

        fountains = features.compactMap { feature in
            guard let geometry = feature["geometry"] as? [String: Any],
                  let coords = geometry["coordinates"] as? [Double], coords.count >= 2,
                  let props = feature["properties"] as? [String: Any],
                  let id = props["objectid"] as? Int else { return nil }

            let wasserart = (props["wasserart"] as? String) ?? ""
            let isTrinkwasser = !wasserart.isEmpty &&
                wasserart.lowercased() != "kein trinkwasser"

            let baujahr: Int? = {
                if let h = props["historisches_baujahr"] as? Int { return h }
                return props["baujahr"] as? Int
            }()

            return Fountain(
                id: id,
                name: (props["standort"] as? String) ?? "Unbekannter Brunnen",
                location: props["ortsbezeichnung"] as? String,
                quartier: props["quartier"] as? String,
                baujahr: baujahr,
                brunnenNummer: props["brunnennummer"] as? String,
                wasserart: wasserart.isEmpty ? nil : wasserart,
                isTrinkwasser: isTrinkwasser,
                latitude: coords[1],
                longitude: coords[0],
                photoUrl: (props["foto"] as? String).flatMap { $0.isEmpty ? nil : $0 },
                isAbgestellt: ((props["abgestellt"] as? String) ?? "nein").lowercased() == "ja"
            )
        }
    }

    func sortedByDistance(from location: CLLocation) -> [(fountain: Fountain, distance: CLLocationDistance)] {
        fountains
            .map { ($0, location.distance(from: $0.clLocation)) }
            .sorted { $0.1 < $1.1 }
    }
}
